File: backend/SalaryCalc/SalaryCalcAPI/reuseable/my_crypto.py
import string
import random

from .SECRET import SECRET_TOKEN, CHAR_LIST

secret = SECRET_TOKEN


def code_message(username, message, operation):
    new_message = ""
    transition = len(username)//3
    char_list = CHAR_LIST
    char_list_length = len(char_list)
    final_transition = 0
    counter = 0
    if(operation!='code'):
        message=message[:len(message)//3];
        # print('message!!!'+message)
    for x in message:
        while counter > len(secret) - 1:
            counter = counter - len(secret)
        # print('x:  '+x )
        final_transition = secret[counter] + transition

        if operation == "code":
            converted = move_forward(
                char_list.index(x), final_transition, char_list_length - 1
            )
        else:
            converted = move_backward(
                char_list.index(x), final_transition, char_list_length - 1
            )
        # print('afterC:  '+char_list[converted])
        new_message += char_list[converted]
    if(operation=='code'):
        for x in range(0, len(message)*2):
            new_message+=char_list[random.randint(0, char_list_length-1)]
    # print('final'+new_message)

    return new_message


def move_forward(start, transition, max_idx):
    start += transition
    while start > max_idx:
        start = start - max_idx
    return start


def move_backward(start, transition, max_idx):
    start -= transition
    while start < 0:
        start = start + max_idx
    return start


print(code_message('admin','admin','code'))
print(code_message('admin','2629383439','decode'))

File: frontend/src/app/reuseable/myCrypto.ts
import { CHAR_LIST, SECRET_TOKEN } from 'src/app/reuseable/SECRET'

export function code(username: string, message, operation) {
    let newMessage = '';
    let transition = username.length;
    let charList = CHAR_LIST
    let secretLength = SECRET_TOKEN.length;
    let charListLength = charList.length;
    let finalTransition = 0;
    let counter = 0;
    let converted: number;

    if(operation!='code'){
        message=message.slice(0, message.length/3)
    }

    for (let i = 0; i < message.length; i++) {
        counter = checkCounter(counter, secretLength);
        finalTransition = SECRET_TOKEN[counter] + transition;
        if (operation == 'code') {
            converted = moveForward(charList.indexOf(message[i]), finalTransition, charListLength)
        } else {
            converted = moveBackward(charList.indexOf(message[i]), finalTransition, charListLength)
        }
        newMessage += charList[converted];
    }
    let extrasSize = message.length * 2;
    if (operation == 'code') {
        for (let i = 0; i < extrasSize; i++) {
            newMessage += charList[Math.floor(Math.random() * 9)]
        }
    }




    // message.map(letter => {
    //     counter = checkCounter(counter, secretLength);
    //     finalTransition = SECRET_TOKEN[counter] + transition;
    //     if (operation == 'code') {
    //         converted = moveForward(charList.indexOf(letter), finalTransition, charListLength)
    //     }else{
    //         converted=moveBackward(charList.indexOf(letter), finalTransition, charListLength)
    //     }
    //     newMessage+=converted;
    // });
    // console.log('newMessage'+newMessage);
    return newMessage;
}

function checkCounter(counter: number, secretLegth: number) {
    while (counter > (secretLegth - 1)) {
        counter = counter - secretLegth;
    }
    return counter;
}

function moveForward(start, transition, max_idx) {
    start += transition
    while (start > max_idx) {
        start = start - max_idx
    }
    return start
}


function moveBackward(start, transition, max_idx) {
    start -= transition
    while (start < 0) {
        start = start + max_idx
    }
    return start
}